python/cudf/cudf/dataframe/multiindex.py

```diff
@@ -375,6 +375,39 @@ def take(self, indices):
         result.names = self.names
         return result
 
+    def serialize(self):
+        """Serialize into pickle format suitable for file storage or network
+        transmission.
+        """
+        header = {}
+        header["type"] = pickle.dumps(type(self))
+        header["names"] = pickle.dumps(self.names)
+        header["codes"], codes_frame = self.codes.serialize()
+        level_frames = []
+        for l in self.levels:
+            header["levels"], f = l.serialize()
+            level_frames.append(f)
+
+        frames = [codes_frame] + level_frames
+        return header, frames
```

Comment:
We don't want to operate with levels and codes here, they are lazily computed attributes solely there for Pandas compatibility. We want to use `_source_data` instead which is a DataFrame of columns that can be (de)serialized normally.

Reply:
If we want to de(serialize) the codes / levels if they've been computed then we should access the underlying attributes directly and check if they're define to prevent this code from unnecessarily materializing them.

Reply:
👍 fixed